cfd_protocol/src/oracle.rs

```diff
@@ -1,75 +1,32 @@
 pub use secp256k1_zkp::*;
 
-use bdk::bitcoin::hashes::Hash;
-use bip340_hash::Bip340Hash;
-use rand::{CryptoRng, RngCore};
-use secp256k1_zkp::bitcoin_hashes::sha256;
-use secp256k1_zkp::{schnorrsig, SecretKey};
-
-mod bip340_hash;
-mod secp_utils;
-
-/// Sign `msg` with the oracle's `key_pair` and a pre-computed `nonce`
-/// whose corresponding public key was included in a previous
-/// announcement.
-pub fn attest(
-    key_pair: &schnorrsig::KeyPair,
-    nonce: &SecretKey,
-    msg: &[u8],
-) -> schnorrsig::Signature {
-    let msg = secp256k1_zkp::Message::from_hashed_data::<sha256::Hash>(msg);
-    secp_utils::schnorr_sign_with_nonce(&msg, key_pair, nonce)
-}
-
-pub fn nonce(rng: &mut (impl RngCore + CryptoRng)) -> (SecretKey, schnorrsig::PublicKey) {
-    let nonce = SecretKey::new(rng);
-
-    let key_pair = schnorrsig::KeyPair::from_secret_key(SECP256K1, nonce);
-    let nonce_pk = schnorrsig::PublicKey::from_keypair(SECP256K1, &key_pair);
-
-    (nonce, nonce_pk)
-}
-
-pub fn msg_hash(
-    pk: &schnorrsig::PublicKey,
+use anyhow::Result;
+use secp256k1_zkp::schnorrsig;
+use std::num::NonZeroU8;
+
+/// Compute an attestation public key for the given oracle public key,
+/// announcement nonce public key and outcome index.
+pub fn attestation_pk(
+    oracle_pk: &schnorrsig::PublicKey,
     nonce_pk: &schnorrsig::PublicKey,
-    msg: &[u8],
-) -> Vec<u8> {
-    let mut buf = Vec::<u8>::new();
-    buf.extend(&nonce_pk.serialize());
-    buf.extend(&pk.serialize());
-    buf.extend(
-        secp256k1_zkp::Message::from_hashed_data::<sha256::Hash>(msg)
-            .as_ref()
-            .to_vec(),
-    );
-
-    Bip340Hash::hash(&buf).into_inner().to_vec()
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use rand::thread_rng;
+    index: NonZeroU8,
+) -> Result<secp256k1_zkp::PublicKey> {
+    let nonce_pk = schnorr_pubkey_to_pubkey(nonce_pk);
+    let nonce_pk_sum = (0..index.get()).map(|_| &nonce_pk).collect::<Vec<_>>();
 
-    fn verify(sig: &schnorrsig::Signature, msg: &[u8], pk: &schnorrsig::PublicKey) -> bool {
-        let msg = secp256k1_zkp::Message::from_hashed_data::<sha256::Hash>(msg);
-        SECP256K1.schnorrsig_verify(sig, &msg, pk).is_ok()
-    }
+    let nonce_pk_sum = PublicKey::combine_keys(&nonce_pk_sum)?;
```

Comment:
I think this could also be solved by https://docs.rs/secp256k1/0.20.3/secp256k1/key/struct.PublicKey.html#method.mul_assign, not sure what is more convenient.